zarr/core.py

```diff
@@ -1821,7 +1821,7 @@ def _encode_chunk(self, chunk):
             cdata = chunk
 
         # ensure in-memory data is immutable and easy to compare
-        if isinstance(self.chunk_store, dict):
+        if isinstance(self.chunk_store, MutableMapping):
             cdata = ensure_bytes(cdata)
```

Comment:
I'm not sure why we don't want to always do that; there is the question of the Consolidated metadata one where we might not want to serialize all, and this might also trigger ensure_bytes  for more stores. I came across this why trying to make a base class for all stores and have store for example always have a `close()` method, so I wrapped `dict()` in a memorystore-like object and test were failing.

Reply:
This is only intended for in-memory stores. In that case we coerce to `bytes` because they are immutable.

We don't want to do this generally as this incurs a copy and if we are just going to write to disk or send over the wire, there is no need.

Reply:
I understand that, the problem is  tell stores implementations they may not receive bytes, and have to deal with it, which is weird as now the stores `__getitem__` might not return the same as `__setitem__`. But fine let's assume we are ok with this. 

The following (existing) test fails for MutableMapping that are not dict, and do not call something  like `ensure_bytes`.

```
    def test_store_has_bytes_values(self):
        # Test that many stores do hold bytes values.
        # Though this is not a strict requirement.
        # Should be disabled by any stores that fail this as needed.

        # Initialize array
        np.random.seed(42)
        z = self.create_array(shape=(1050,), chunks=100, dtype='f8', compressor=[])
        z[:] = np.random.random(z.shape)

        # Check in-memory array only contains `bytes`
        assert all([isinstance(v, bytes) for v in z.chunk_store.values()])

        if hasattr(z.store, 'close'):
            z.store.close()
```

And note that the name of the test is xxxx_store_xxxx, not xxx_dict_xxxx

So IMHO either we:
 - stick to stores have bytes interfaces, 
 - relax stores interface and tell them "you may get things that are not bytes", deal with it.
 - have a proper way to have a store tell "I'm ok not receiving bytes". 

Otherwise many of the tests rely on all stores having all the same behavior but testing only with `dict()` which is special cased.

Regardless I'll revert that and we can move to to another issue.